### ISSUE TITLE
Add brotli compression for data uploading

### ISSUE THREAD

**Issue by mboecker** [mboecker/mlrmbo-webservice#7]
Brotli can decrease the data size by about 50%, we can do this in both the client ("brotli") and the server (python brotli binding). As brotli is used by dropbox and others, its performance is actually optimized for this use case.

**Comment by jakob-r** [comment]
I think this is not so relevant because we are not compressing text or html but just (mostly non repeating) numbers and therefore a dictionary (especially a predefined one for text and html) does not help much?

**Comment by jakob-r** [comment]
also the data that is transfered is already quite small, right?

**Comment by mboecker** [comment]
I was thinking we could use this for the initial data upload by the user, which can be arbitrary large. Using a compression method here (be it gzip or brotli) would lower the number of requests necessary, because we can only submit about 1000 data points per query.

Data compression can reduce the size of our query, because we're sending ASCII-encoded strings, containing the numbers. Any compression method will therefore reduce the size of our string.

If we were only using numeric parameters, we could simply send the data binary-encoded, which would probably the most efficient and effective way to "compress" them. But since we're also transmitting factorial parameters, I chose to use ascii-string-compression instead.

I chose brotli for this, because I heard of it recently and thought it is the new and best compression method availiable: "Brotli decompression is at least as fast as for gzip while significantly improving the compression ratio."

But: "The price we pay is that compression is much slower than gzip."

(source)[https://www.opencpu.org/posts/brotli-benchmarks]

I didn't know about that. I think it's still useful to keep compressing data when the user wants to upload more than say 1000 data points. But I think we should be using a different compression method.

**Comment by mboecker** [comment]
I was also thinking, that there are two server endpoints for this /upload and /upload_compressed or something like this. We then could decide on the package side which method to use.

Another thought: We could also use something like BSON, which is basically a binary encoded JSON variant. We could still use strings for the factors but would profit from binary-encoded numbers.

**Comment by jakob-r** [comment]
While everything might be a nice to have they do not contribute so much to the essential feature set of this software. So you can implement all you want but keep in mind that things like documentation and tests etc. are more important for the user in the beginning.
Don't spend to much time on corner cases (aka. 1000 data points in the initial design)

**Comment by mboecker** [comment]
You're right, but this is actually quite easy and mostly done in the brotli-branch.
